validator/index.js

```diff
@@ -25,50 +25,233 @@ const path = require('path');
+     * validator.proto. Examples: 'UNKNOWN_CODE', 'MANDATORY_TAG_MISSING',
+     * 'TAG_REQUIRED_BY_MISSING'. For each of these codes there is a
+     * format string in validator-main.protoascii (look for error_formats),
+     * which is used to assemble the message from the strings in params.
+     * @type {!string}
+     */
+    this.code = 'UNKNOWN_CODE';
+    /**
+     * This field is only useful when scripting against the validator,
+     * it should not be displayed in a user interface as it adds nothing
+     * for humans to read over the message field (see above).
+     * @type {!Array<!string>}
+     */
+    this.params = [];
+  }
+}
+
+/**
+ * A global static map used by the validateString function to avoid loading
+ * AMP Validators more than once.
+ * @type {!Object<string, Object>}
+ */
+const precompiledByValidatorJs = {};
+
+/**
+ * @param {!string} inputString
+ * @param {string=} validatorJs_opt
+ * @returns {Promise<(ValidationResult|Error)>}
+ * @export
+ */
+function validateString(inputString, opt_validatorJs) {
+  const validatorJs =
+      opt_validatorJs || 'https://cdn.ampproject.org/v0/validator.js';
+  // The 'sandbox' is a Javascript object (dictionary) which holds the results
+  // of evaluating the validatorJs, so basically, it holds functions, prototypes,
+  // etc. As a side-effect of evaluating, the VM will compile this code and it's
+  // worth holding onto it. Hence, this validate function is reached via
+  // 2 codepaths - either the sandbox came from the cache,
+  // precompiledByValidatorJs - or we just constructed it after downloading
+  // and evaluating the script. The API is fancier here, vm.Script /
+  // vm.createContext / vm.runInContext and all that, but it's quite similar
+  // to a Javascript eval.
+  const validate = (sandbox, resolve) => {
+    const internalResult = sandbox.amp.validator.validateString(inputString);
+    const result = new ValidationResult();
+    result.status = internalResult.status;
+    for (const internalError of internalResult.errors) {
+      const error = new ValidationError();
+      error.severity = internalError.severity;
+      error.line = internalError.line;
+      error.col = internalError.col;
+      error.message = sandbox.amp.validator.renderErrorMessage(internalError);
+      error.specUrl = internalError.specUrl;
+      error.code = internalError.code;
+      error.params = internalError.params;
+      error.category = sandbox.amp.validator.categorizeError(internalError);
+      result.errors.push(error);
     }
+    resolve(result);
+  };
+  if (precompiledByValidatorJs.hasOwnProperty(validatorJs)) {
+    return new Promise(function(resolve, reject) {
+      const sandbox = precompiledByValidatorJs[validatorJs];
+      validate(sandbox, resolve);
+    });
+  }
+  const validatorJsPromise =
+      (isHttpOrHttpsUrl(validatorJs) ? readFromUrl : readFromFile)(validatorJs);
+  return new Promise(function(resolve, reject) {
```

Comment:
You can chain of the promise instead of using the constructor:

```js
return validatorJsPromise.then((scriptContents) => {
  const script = new vm.Script(scriptContents);
  const sandbox = vm.createContext();
  script.runInContext(sandbox);
  precompiledByValidatorJs[validatorJs] = sandbox;
  return validate(sandbox);
});